IOMC/ParticleGuns/src/CloseByParticleGunProducer.cc

```diff
@@ -0,0 +1,125 @@
+#include <ostream>
+
+#include "IOMC/ParticleGuns/interface/CloseByParticleGunProducer.h"
+
+#include "SimDataFormats/GeneratorProducts/interface/HepMCProduct.h"
+#include "SimDataFormats/GeneratorProducts/interface/GenEventInfoProduct.h"
+
+#include "DataFormats/Math/interface/Vector3D.h"
+
+#include "FWCore/Framework/interface/Event.h"
+#include "FWCore/ParameterSet/interface/ParameterSet.h"
+#include "FWCore/ServiceRegistry/interface/Service.h"
+#include "FWCore/Utilities/interface/RandomNumberGenerator.h"
+#include "FWCore/MessageLogger/interface/MessageLogger.h"
+
+#include "CLHEP/Random/RandFlat.h"
+#include "CLHEP/Units/GlobalSystemOfUnits.h"
+#include "CLHEP/Units/GlobalPhysicalConstants.h"
+#include "CLHEP/Random/RandFlat.h"
+
+using namespace edm;
+using namespace std;
+
+CloseByParticleGunProducer::CloseByParticleGunProducer(const ParameterSet& pset) :
+   BaseFlatGunProducer(pset)
+{
+
+  ParameterSet defpset ;
+  ParameterSet pgun_params =
+    pset.getParameter<ParameterSet>("PGunParameters") ;
+
+  fEn = pgun_params.getParameter<double>("En");
+  fR = pgun_params.getParameter<double>("R");
+  fZ = pgun_params.getParameter<double>("Z");
+  fDelta = pgun_params.getParameter<double>("Delta");
+  fPartIDs = pgun_params.getParameter< vector<int> >("PartID");
+  fPointing = pgun_params.getParameter<bool>("Pointing");
+
+  produces<HepMCProduct>("unsmeared");
+  produces<GenEventInfoProduct>();
+}
+
+CloseByParticleGunProducer::~CloseByParticleGunProducer()
+{
+   // no need to cleanup GenEvent memory - done in HepMCProduct
+}
+
+void CloseByParticleGunProducer::produce(Event &e, const EventSetup& es)
+{
+   edm::Service<edm::RandomNumberGenerator> rng;
+   CLHEP::HepRandomEngine* engine = &rng->getEngine(e.streamID());
+
+   if ( fVerbosity > 0 )
+     {
+       LogDebug("CloseByParticleGunProducer") << " CloseByParticleGunProducer : Begin New Event Generation" << endl ;
+     }
+   fEvt = new HepMC::GenEvent() ;
+
+   // loop over particles
+   //
+   int barcode = 1 ;
+   double phi = CLHEP::RandFlat::shoot(engine, -3.14159265358979323846, 3.14159265358979323846);
+   for (unsigned int ip=0; ip<fPartIDs.size(); ++ip, phi += fDelta/fR)
+   {
+
+     int PartID = fPartIDs[ip] ;
+     const HepPDT::ParticleData *PData = fPDGTable->particle(HepPDT::ParticleID(abs(PartID))) ;
+     double mass   = PData->mass().value() ;
+     double mom    = sqrt(fEn*fEn-mass*mass);
+     double px     = 0.;
+     double py     = 0.;
+     double pz     = mom;
+     double energy = fEn;
+
+     // Compute Vertex Position
+     double x=fR*cos(phi);
+     double y=fR*sin(phi);
+     constexpr double c= 2.99792458e+1; // cm/ns
+     double timeOffset = sqrt(x*x + y*y + fZ*fZ)/c*ns*c_light;
```

Comment:
@rovere , I would expect here CMS units (cm, GeV, ns)? In that case of CLHEP (mm, MeV, ns) should not be used.
Having in one expression both your constant "c" and CLHEP "c_light" seems to be a problem.

To resolve this I would put a comment line about fX, fX, fZ unit, and select one or another.

Reply:
I used CMS units for the configuration parameters, namely cm, GeV and ns. Then I compute the timing in ns using CMS units and the "correct" c value and convert it back to whatever CLHEP/hepmc is expecting. Finally, I put back also x,y,fZ back to units that are digested by HepMC.
I have to admit that this mix of units is, to put it mildly, quite ugly and confusing, I agree, but it's not coming from this code but rather imposed from externals.
By the way, I think you will find the same kind of "units massaging" also in other Generators.